### PR TITLE
html: Update tests for an autofocus algorithm update

### PR DESCRIPTION
**Pull request by tkent-google**
https://github.com/whatwg/html/pull/4763

### CODE REVIEW

File: html/semantics/forms/autofocus/focusable-area-in-top-document.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+<script>
+'use strict';
+
+promise_test(async () => {
+  await waitForLoad(window);
+  let iframe = document.querySelector('iframe');
+  assert_equals(document.activeElement, iframe);
+
+  let input = document.createElement('input');
+  input.autocous = true;
```
Comment (domenic) [inline html/semantics/forms/autofocus/focusable-area-in-top-document.html]:
```suggestion
  input.autofocus = true;
```

File: html/semantics/forms/autofocus/skip-document-with-fragment.html [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  assert_equals(document.activeElement, document.body,
+      'Autofocus elements in iframes should not be focused.');
+
+  let input = document.createElement('input');
+  input.autofocus = true;
+  document.body.appendChild(input);
+  await waitForAnimationFrame();
+  await waitForAnimationFrame();
+  assert_equals(document.activeElement, input);
+}, 'Autofocus element in a document with a URI fragment should be skipped.');
```
Comment (domenic) [inline html/semantics/forms/autofocus/skip-document-with-fragment.html]:
```suggestion
}, 'Autofocus elements in iframed documents with URL fragments should be skipped.');
```

**Review by domenic — CHANGES_REQUESTED** [review]
Found a typo bug that needs fixing.

Things I see in the spec but I am not sure if there are tests for:

- Sandboxing should prevent autofocus
- Cross-origin should prevent autofocus
- It should wait for script-blocking style sheets, possibly for a long time. (Best done with a slow server response for a stylesheet.)
- If topDocument's URL fragment is not empty, should not autofocus. (Can test using window.open() to create a new top-level BC)
- Even with no blocking style sheets, if you wait a long time (~100 ms?), and the focus is still on the document itself, inserting an autofocus element will cause that element to be focused after the next frame.

Also worth noting that we should test the exact position in the event loop's "update the rendering" for when autofocusing happens. I'm not sure the best way to do that with the current spec proposal, although I think your use of double-rAF instead of single-rAF is probably because of the fact that it's after rAF. But for example, if we stick with the current spec proposal, we should test that after a single rAF or a single IntersectionObserver callback, there is no autofocus, but before the second resize event, the autofocus does happen.

I don't think it's necessary to land all these tests, but given how historically non-interoperable this attribute is, the more tests we can land, the better.

File: html/semantics/forms/autofocus/focusable-area-in-top-document.html (hunk 1 above)
Comment (tkent-google) [reply html/semantics/forms/autofocus/focusable-area-in-top-document.html]: Done.

File: html/semantics/forms/autofocus/skip-document-with-fragment.html (hunk 2 above)
Comment (tkent-google) [reply html/semantics/forms/autofocus/skip-document-with-fragment.html]: Done.

**Comment by tkent-google** [comment]
> Sandboxing should prevent autofocus

Added no-sandboxed-automatic-features.html.

> Cross-origin should prevent autofocus

We already have no-cross-origin-autofocus.sub.html.

> It should wait for script-blocking style sheets, possibly for a long time. (Best done with a slow server response for a stylesheet.)

Added spin-by-blocking-style-sheet.html.

> If topDocument's URL fragment is not empty, should not autofocus. (Can test using window.open() to create a new top-level BC)

Added new test case to skip-document-with-fragment.html.

> Even with no blocking style sheets, if you wait a long time (~100 ms?), and the focus is still on the document itself, inserting an autofocus element will cause that element to be focused after the next frame.

Added autofocus-on-stable-document.html.

> Also worth noting that we should test the exact position in the event loop's "update the rendering" for when autofocusing happens. ...

Added update-the-rendering.html.

**Comment by tkent-google** [comment]
update-the-rendering.html is flaky with Firefox because the order of an animation frame callback and a user interaction task, in which autofocus processing is currently done, is not deterministic.  I have no idea how to fix this issue, and might remove the test.

**Review by domenic — APPROVED** [review]
If Firefox implemented the proposed spec, then the test would not be flaky, right? In that case I think we should keep the test, and ask the admins for an exception to allow merging this.

**Comment by tkent-google** [comment]
> If Firefox implemented the proposed spec, then the test would not be flaky, right?

That's right.

**Comment by foolip** [comment]
Should this be merged now? I can admin merge if needed.

**Comment by domenic** [comment]
@foolip yes, please admin merge!

**Comment by foolip** [comment]
Done!

**Pull request merged**